### PR TITLE
Proposal for merge: Mapsforge 6 as experimental map

### PR DESCRIPTION
**Pull request by rsudev**
I think I have stabilised the current implementation sufficiently and at the same time minimised the impact on the existing functionality to propose this as an experimental feature for merging.
It is based on mapsforge 0.6.0, the latest released version. It works with the latest, SVG-based map styles.

This implementation can under the extremely creative title 'New Map' be used as a normal navigation target. Additionally can it be defined as replacement for the live- and list-maps in Settings - System.

Functionality-wise does it mainly cover everything the current map does besides some unimplemented menu items (e.g. store offline) and an online mode.

Additional functionality:
* Selection menu when tapping a place with several caches
* Switch navigation target (red line) in popup
* Long-press on the map to add a waypoint there to the current target cache

Technical notes:
* I recompiled the current mapsforge library (0.3) in a different package, to avoid clashes with the new version (0.6)
* In this implementation I did not upgrade to appcompat 23 first, to avoid degrading the experience with the current map

Depending on feedback I would from my side propose to merge in about two weeks, to give everybody enough time to take a look on this not so small changeset...

---
This change is [Reviewable](https://reviewable.io/reviews/cgeo/cgeo/5584)

### CODE REVIEW

**Comment by SammysHP** [comment]
Great! Does this already support style options (required by the new Freizeitkarte theme)?

**Comment by rsudev** [comment]
I would assume that, but didn't check it. I currently use openandromaps and the ElevateSVG-theme...

**Comment by SammysHP** [comment]
With "style options" I mean the ability to configure a selected theme (enable/disable features).

**Comment by rsudev** [comment]
Nothing specifically implemented. Do you have some pointers?

**Comment by samueltardieu** [comment]
@UniQP Please open a separate issue if you want to discuss this style point. We do *not* require a newline at end of file.

**Comment by Lineflyer** [comment]
> This implementation can under the extremely creative title 'New Map' be used as a normal navigation target. Additionally can it be defined as replacement for the live- and list-maps in Settings - System.

I guess, you did it this way because of the experimental status or shall this stay like this in the final implementation?

Besides that, I can offer to test this implementation after merging. I can not judge the code, but to be merged it should be in a shape that it is realistic to include it in the next feature release (so maybe 4-6 weeks from now typically).

**Comment by rsudev** [comment]
In my opinion it can be included in a release, as an experimental feature.
My plan would be, while maintaining/polishing this implementation to start to work on moving Google maps to a similar/integrated implementation, based on gmaps v2.
Once this is ready as well, we can make the 'new map' the only map again.
I could continue to do this on my private branch, but it is easier to have that in master.
To put it into perspective: This implementation comprises the larger part of my c:geo development work for the past ~ 18 month, so if I can complete the second part in a shorter time frame (6 - 12 month), would of course be good.
Currently I see no other way to get a map migration moving.

**Comment by samueltardieu** [comment]
*[main/build.gradle, line 234 \[r1\]](https://reviewable.io:443/reviews/cgeo/cgeo/5584#-KEXMAkFcvqEAs_VkajZ:-KEXMAkFcvqEAs_Vkaj_:632400094) ([raw file](https://github.com/cgeo/cgeo/blob/96eb6b08b537005253c32b0d16c4cba5a26b6334/main/build.gradle#L234)):*
Can't you add them with the regular gradle mechanism which automatically fetches the files?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cgeo/cgeo/5584)*

**Comment by Bananeweizen** [comment]
I'm in favor of merging and afterwards massaging it. 4 KLOC are quite hard to review and need to be played with. If there are problems, we can still completely hide the feature for the next release.

File: main/src/cgeo/geocaching/maps/mapsforge/v6/caches/AbstractCachesOverlay.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        final Geopoint target = waypoint.getCoords();
+
+        if (target != null) {
+            final Bitmap marker = AndroidGraphicFactory.convertToBitmap(MapUtils.getWaypointMarker(CgeoApplication.getInstance().getResources(), waypoint));
+            return new GeoitemLayer(waypoint.getGeoitemRef(), tapHandler, new LatLong(target.getLatitude(), target.getLongitude()), marker, 0, -marker.getHeight() / 2);
+        }
+
+        return null;
+    }
+ }
```
Comment (rsudev) [inline main/src/cgeo/geocaching/maps/mapsforge/v6/caches/AbstractCachesOverlay.java]: I can do that

**Comment by rsudev** [comment]
Review status: 0 of 88 files reviewed at latest revision, 2 unresolved discussions.

---

*[main/build.gradle, line 234 \[r1\]](https://reviewable.io:443/reviews/cgeo/cgeo/5584#-KEXMAkFcvqEAs_VkajZ:-KEeykXwksMlJzKdKKqo:1887948611) ([raw file](https://github.com/cgeo/cgeo/blob/96eb6b08b537005253c32b0d16c4cba5a26b6334/main/build.gradle#L234)):*
When I started, gradle was not even on the horizon for us :)
Nevertheless I tried right away, but there seems to be an issue with getting androidsvg, so I'd rather postpone this step.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cgeo/cgeo/5584)*

File: main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java [hunk 2]
```diff
@@ -0,0 +1,1217 @@
...
+
+            } else {
+                showPopup(items.get(0));
+            }
+
+        } catch (final NotFoundException e) {
+            Log.e("NewMap.showPopup", e);
+        }
+
+        return;
```
Comment (kumy) [inline main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java]: Codacy Issue found: [Avoid unnecessary return statements](https://www.codacy.com/app/cgeo/cgeo/file/2364754834/issues/source?bid=3217125&fileBranchId=3217166#l1089)

File: main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java [hunk 3]
```diff
@@ -0,0 +1,1217 @@
...
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import rx.Subscription;
+import rx.subscriptions.Subscriptions;
+
+@SuppressLint("ClickableViewAccessibility")
+public class NewMap extends AbstractActionBarActivity {
```
Comment (kumy) [inline main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java]: Codacy Issue found: [Avoid really long classes.](https://www.codacy.com/app/cgeo/cgeo/file/2364754834/issues/source?bid=3217125&fileBranchId=3217166#l91)

File: main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java [hunk 4]
```diff
@@ -0,0 +1,1215 @@
...
```
Comment (kumy) [inline main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java]: Codacy Issue found: [Avoid really long classes.](https://www.codacy.com/app/cgeo/cgeo/file/2364815869/issues/source?bid=3217125&fileBranchId=3217166#l91)

Comment (rsudev) [inline main/src/cgeo/geocaching/maps/mapsforge/v6/NewMap.java]: That would be one of the topics of restructuring the current implementation after merging...

**Comment by rsudev** [comment]
If there are no objections, I will rebase again and push this on the weekend.

**Comment by Bananeweizen** [comment]
Sounds fine to me.

**Comment by Bananeweizen** [comment]
Any more work necessary? Otherwise I would simply merge as is...

**Comment by rsudev** [comment]
I will rebase once more and push directly, I'd say

**Comment by rsudev** [comment]
Pushed directly.

**Pull request closed without merging**